azure-iot-device/azure/iot/device/provisioning/sk_provisioning_device_client.py

```diff
@@ -0,0 +1,87 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+This module contains one of the implementations of the Provisioning Device Client which uses Symmetric Key authentication.
+"""
+import logging
+from threading import Event
+import traceback
+from .provisioning_device_client import ProvisioningDeviceClient
+from .internal.polling_machine import PollingMachine
+
+logger = logging.getLogger(__name__)
+
+
+class SymmetricKeyProvisioningDeviceClient(ProvisioningDeviceClient):
+    """
+    Client which can be used to run the registration of a device with provisioning service
+    using Symmetric Key authentication.
+    """
+
+    def __init__(self, mqtt_state_based_provider):
+        """
+        Initializer for the Symmetric Key Registration Client
+        :param mqtt_state_based_provider: The state-based protocol provider. As of now this only supports MQTT.
+        """
+        super(SymmetricKeyProvisioningDeviceClient, self).__init__(mqtt_state_based_provider)
+        self._polling_machine = PollingMachine(mqtt_state_based_provider)
+        # To be defined by sample
+        self.on_registration_complete = None
+        self._polling_machine.on_disconnected = self._on_connection_state_change
+        self._polling_machine.on_registration_complete = self._on_device_registration_complete
+
+    def register(self):
+        """
+        Register the device with the provisioning service.
+        """
+        logger.info("Registering with Hub...")
+        register_complete = Event()
+
+        def callback_register(result=None, error=None):
+            # This could be a failed/successful registration result from the HUB
+            # or a error from polling machine. Response should be given appropriately
+            if result is not None:
+                if result.status == "assigned":
+                    logger.info("Successfully registered with Hub")
+                else:  # There be other statuses
+                    logger.error("Failed registering with Hub")
+            if error is not None:  # This can only happen when the polling machine runs into error
+                logger.info(error)
+
+            register_complete.set()
+
+        self._polling_machine.register(callback=callback_register)
+
+        register_complete.wait()
+
+    def cancel(self):
+        """
+        Cancels the current registration process.
+        """
+        logger.info("Cancelling the current registration process")
+        cancel_complete = Event()
+
+        def callback_cancel():
```

Comment:
>callback_cancel [](start = 12, length = 15)

I've been pushing for, as a convention, callbacks to not be referred to as a callback until they are given to the thing that calls back. i.e. this function shouldn't be called callback, it should be called on_cancel, and passed in the "callback" parameter to polling_machine.cancel.

The idea is that the word "callback" exclusively refers to a function given to you, to call back, not a function you define to give to someone else to call back with.

i.e. the term "callback" is only used by something that consumes a callback, not provides one.

The rationale is that as we go down the stack and there are many layers of callbacks, it becomes confusing as to which you're referring to with the term "callback" - is it a function you are providing to another layer to call back to you with, or is it a function you are consuming that you will call back on?

The goal is to start only using the term "callback" for the latter case. Bert and I have been moving this way for the past few weeks in the transport refactor, and while not all code has been updated under this model, it's an ongoing effort.

All new code should adopt this convention going forth. Ask Bert about it if this is confusing at all.

This is a "here and everywhere" type comment of course, but i'll try to flag any that jump out at me in separate comments.

Reply:
I like it to be called on_complete_cancel or on_cancel_complete 
The only reason I never called it the cancel_complete coz that was the event name as well.

However I still see that sync clients have the name as callback in the same pattern .

---
In reply to: [282629484](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282629484) [](ancestors = 282629484)

Reply:
Yeah the sync_clients haven't been updated to conform yet.

Here's the document I promised:
[https://github.com/Azure/azure-iot-sdk-python-preview/wiki/callbacks](https://github.com/Azure/azure-iot-sdk-python-preview/wiki/callbacks)

---
In reply to: [282637616](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282637616) [](ancestors = 282637616,282629484)